### PR TITLE
Skip clusterdeploy handler if downstream cluster is disconnected

### PR DESCRIPTION
**Pull request by moio**
## Problem
With a disconnected downstream cluster, the `clusterdeploy` handler takes several seconds to time out. This is fine with only a few downstream clusters disconnected, but when the number of disconnected clusters goes over 50 or so, the worker queues will get saturated.

Reason for the slow time out is that the handler may execute code that reaches out to the disconnected downstream cluster - typically `clusterDeploy.getNodeAgentImage` which tries to list `DaemonSets`. That alone can take up to ~30 because `TunnelServer`'s `remotedialer` mechanism retries on error.
 
## Solution
Exploit the fact that the `clusterconnected` controller is already marking the cluster status as disconnected. Use that status to exit `clusterdeploy` early in presence of disconnected clusters. Nothing on disconnected clusters can be changed anyway, and trying to read from them is going to fail.~

Make `TunnelServer`'s `remotedialer` mechanism retrying optional. Avoid retrying in the `clusterdeploy` handler - so that code fails fast.

## Testing
See [reproduction steps on the issue](https://github.com/rancher/rancher/issues/39597#issuecomment-1816099261).

## Engineering Testing
Per reproduction steps above (by @moio).

### Manual Testing

~This was tested in our production cluster which has a lot (hundreds+) of disconnected clusters and all work queues were blocked. After the patch `cluster-deploy` time went from about 15s down to subsecond. Lasso working queue depth went from *approx amount of disconnected clusters* to 0-1-2.~ needs re-testing

### Automated Testing
* Test types added/modified:
    * Unit: healthsyncer 

Summary: to be discussed whether it is sufficient

## QA Testing Considerations

No specific recommendations on fresh and upgrade scenarios.
 
### Regressions Considerations

Clusters should continue to be displayed as available or not available in the UI depending on their connection status, with ~1 minute lag, as before this patch. This is covered by the manual tests above.

## Credits

@erkki for isolating problem and first solution attempt.

Supersedes https://github.com/rancher/rancher/pull/43542

### CODE REVIEW

**Comment by moio** [comment]
@snasovich I need you or somebody from your team to take a look at this PR, can you point me to the most appropriate person? Thanks in advance

**Comment by snasovich** [comment]
@moio , assigned @kinarashah @Oats87 @jakefhyde to review. Thank you!

**Comment by erkki** [comment]
Crossref https://github.com/rancher/rancher/issues/39597#issuecomment-1858728220 for potential regression

**Comment by moio** [comment]
> how extensively has this been tested?

@jakefhyde I have been testing this myself, QA instructions are ready.

One customer tested it in a debug image but reported a regression: importing existing AKS clusters would sometimes get stuck and not complete. I tried but failed to reproduce that problem ([details](https://github.com/rancher/rancher/issues/39597#issuecomment-1857824561)).

Another community user [reported a similar problem and I tried to investigate it without coming to a conclusion](https://github.com/rancher/rancher/issues/39597#issuecomment-1858728220).

I would really, really appreciate if you, or anybody else with deeper knowledge, could offer ideas/guidance to move forward and ideally reproduce the problem.

File: pkg/controllers/management/clusterdeploy/clusterdeploy.go [hunk 1]
```diff
@@ -127,6 +127,12 @@ func (cd *clusterDeploy) doSync(cluster *apimgmtv3.Cluster) error {
 		return nil
 	}
 
+	// Skip further work if clusterconnected controller has determined agent not connected
+	if apimgmtv3.ClusterConditionReady.IsFalse(cluster) && apimgmtv3.ClusterConditionReady.GetReason(cluster) == "Disconnected" {
+		logrus.Tracef("clusterDeploy: doSync: cluster [%s] agent is not connected", cluster.Name)
+		return errors.New("cluster agent not connected")
```
[the remark this answers is not in the document]
Comment (moio) [reply pkg/controllers/management/clusterdeploy/clusterdeploy.go]: Addressed as of c3e1b96c2

**Review by Oats87 — CHANGES_REQUESTED** [review]
@moio yoy may need to check to see if the cluster even needs a dialer, see: https://github.com/rancher/rancher/blob/5314106f746c0c62e95d54922a2dfd116d2f6f68/pkg/dialer/factory.go#L189-L192 

there are specific cluster types where we don't dial using the tunnel dialer

**Comment by jakefhyde** [comment]
Killed the build since it was stuck waiting for s390x.

**Comment by moio** [comment]
> @moio yoy may need to check to see if the cluster even needs a dialer, see:
> 
> https://github.com/rancher/rancher/blob/5314106f746c0c62e95d54922a2dfd116d2f6f68/pkg/dialer/factory.go#L189-L192
> 
> there are specific cluster types where we don't dial using the tunnel dialer

I did not know about that part of the code, I learned something, thanks for sharing.

Nevertheless I fail to see how the fact of using a tunnel dialer or not could impact the proposed change in this PR.

Please correct any errors in my reasoning:

1. this patch relies on the `ClusterConditionReady` condition being correctly set
2. the only place in the code where that is set is in [HealthSyncer](https://github.com/rancher/rancher/blob/e4e0fdd7991ef236d3d096e6473f42f26393e08d/pkg/controllers/managementuser/healthsyncer/healthsyncer.go#L150-L162), which periodically [attempts to look up a namespace](https://github.com/rancher/rancher/blob/e4e0fdd7991ef236d3d096e6473f42f26393e08d/pkg/controllers/managementuser/healthsyncer/healthsyncer.go#L89) as a means to check whether the cluster's API is ready
3. that namespace look up [uses the `HealthSyncer.k8s`](https://github.com/rancher/rancher/blob/e4e0fdd7991ef236d3d096e6473f42f26393e08d/pkg/controllers/managementuser/healthsyncer/healthsyncer.go#L67) client which comes from the `UserContext` used at registration time
4. that `UserContext` in turn [comes from the Cluster Manager](https://github.com/rancher/rancher/blob/3c1c5730afc57546b362b2010744efdc7b77212b/pkg/clustermanager/manager.go#L223)'s internal record
5. that is [in turn configured by ToRESTConfig](https://github.com/rancher/rancher/blob/3c1c5730afc57546b362b2010744efdc7b77212b/pkg/clustermanager/manager.go#L414-L425) which, crucially, calls the Dialer factory's `ClusterDialer` func
6. `ClusterDialer` further delegates to `clusterFunc` which is where the code you pointed at decides if a tunnel is needed or not

Now no matter how `clusterFunc` decides, I expect to see an endpoint that returns namespaces (possibly 0) if the API works.

Are there reasons I am missing why the choice of whether using the tunnel or not could have effects?

Thanks in advance!

**Comment by Oats87** [comment]
> > @moio yoy may need to check to see if the cluster even needs a dialer, see:
> > https://github.com/rancher/rancher/blob/5314106f746c0c62e95d54922a2dfd116d2f6f68/pkg/dialer/factory.go#L189-L192
> > 
> > there are specific cluster types where we don't dial using the tunnel dialer
> 
> I did not know about that part of the code, I learned something, thanks for sharing.
> 
> Nevertheless I fail to see how the fact of using a tunnel dialer or not could impact the proposed change in this PR.
> 
> Please correct any errors in my reasoning:
> 
>     1. this patch relies on the `ClusterConditionReady` condition being correctly set
> 
>     2. the only place in the code where that is set is in [HealthSyncer](https://github.com/rancher/rancher/blob/e4e0fdd7991ef236d3d096e6473f42f26393e08d/pkg/controllers/managementuser/healthsyncer/healthsyncer.go#L150-L162), which periodically [attempts to look up a namespace](https://github.com/rancher/rancher/blob/e4e0fdd7991ef236d3d096e6473f42f26393e08d/pkg/controllers/managementuser/healthsyncer/healthsyncer.go#L89) as a means to check whether the cluster's API is ready
> 
>     3. that namespace look up [uses the `HealthSyncer.k8s`](https://github.com/rancher/rancher/blob/e4e0fdd7991ef236d3d096e6473f42f26393e08d/pkg/controllers/managementuser/healthsyncer/healthsyncer.go#L67) client which comes from the `UserContext` used at registration time
> 
>     4. that `UserContext` in turn [comes from the Cluster Manager](https://github.com/rancher/rancher/blob/3c1c5730afc57546b362b2010744efdc7b77212b/pkg/clustermanager/manager.go#L223)'s internal record
> 
>     5. that is [in turn configured by ToRESTConfig](https://github.com/rancher/rancher/blob/3c1c5730afc57546b362b2010744efdc7b77212b/pkg/clustermanager/manager.go#L414-L425) which, crucially, calls the Dialer factory's `ClusterDialer` func
> 
>     6. `ClusterDialer` further delegates to `clusterFunc` which is where the code you pointed at decides if a tunnel is needed or not
> 
> 
> Now no matter how `clusterFunc` decides, I expect to see an endpoint that returns namespaces (possibly 0) if the API works.
> 
> Are there reasons I am missing why the choice of whether using the tunnel or not could have effects?
> 
> Thanks in advance!

I may be outdated in my understanding here but the problem is that with older provisioning lifecycles of managed Kubernetes clusters, we relied on Rancher being able to dial the API server of the downstream cluster to initially deploy the `cattle-cluster-agent`. Implementing your proposed fix introduces a chicken-and-egg situation for that case where we are deploy the `cattle-cluster-agent` which runs the `healthsyncer` which means the `Ready` condition will never be set to `True` and we cannot use `cluster-deploy`. 

It's also important to note that we rely on `ClusterDeploy` to "fix" broken configurations of a cluster downstream as well, provided we have some method of dialing into it. This includes the use of the `cattle-node-agent` in RKE1 clusters, which is *not* considered a `cluster-agent` and will not run the controllers necessary.

Thinking about this further I am not necessarily comfortable with our reliance on the "house of cards" you have stated for management of the `cattle-cluster-agent`, rather, I would prefer we find a more robust method of determining whether we have an active dialer connection to the downstream cluster (if one is necessary) and use that to determine whether `clusterdeploy` operates or not.

**Comment by erkki** [comment]
We are currently reverting this in our prod clusters as we see a regression as per https://github.com/rancher/rancher/issues/39597#issuecomment-1858728220 . Need to work on a safe testing environment -- in this case we only saw this regression when over 100 clusters were deployed (with 80% not connected).

**Comment by moio** [comment]
> [...] Implementing your proposed fix [for managed clusters] introduces a chicken-and-egg situation for that case where we are deploy the `cattle-cluster-agent` which runs the `healthsyncer` which means the `Ready` condition will never be set to `True` and we cannot use `cluster-deploy`.

You are right, I could reproduce the problem trying to deploy an AKS cluster. I now understand the `ClusterDialer` factory that is ultimately used by `clusterdeploy` will return a non-tunneled connection to the Kubernetes cluster API which will work before `HealthSync` even starts. I could not see the problem before because all user cases I was studying (and my own setup) were using imported clusters, where this is not an issue.

I now understand and agree this is not the way to go.

> It's also important to note that we rely on `ClusterDeploy` to "fix" broken configurations of a cluster downstream as well, provided we have some method of dialing into it. This includes the use of the `cattle-node-agent` in RKE1 clusters, which is _not_ considered a `cluster-agent` and will not run the controllers necessary.

I now understand that `ClusterDialer` will attempt connecting in various ways, including connecting directly to any control plane node if direct connection is not an option and tunnel is not ready for whatever reason.

> Thinking about this further I am not necessarily comfortable with our reliance on the "house of cards" you have stated for management of the `cattle-cluster-agent`, rather, I would prefer we find a more robust method of determining whether we have an active dialer connection to the downstream cluster (if one is necessary) and use that to determine whether `clusterdeploy` operates or not.

I am working at a new patch that will use the existing mechanisms but the blocking part of connection opening.

I will ask you for another review when that passes my own testing at least.

Thanks for putting me on a better track!

**Comment by moio** [comment]
@Oats87 can you take a look at the new version of this change which is radically different and (to my understanding) follows your comments?

Thanks in advance

**Comment by moio** [comment]
@Oats87 related: what would be your opinion about extending this approach to `statsaggregator` (see 460eb9ac9) and potentially other controllers?

**Review by Oats87 — DISMISSED** [review]
Silvio, this is great.

Thanks for breaking up the changes into "bite size chunks" across commits.

**Review by jakefhyde — APPROVED** [review]
Approving. Left some addressable feedback but honestly it's all minor stuff and pedantic.

File: pkg/clustermanager/manager.go [hunk 2]
```diff
@@ -256,7 +256,11 @@ func (m *Manager) doStart(rec *record, clusterOwner bool) (exit error) {
 	}
 }
 
-func ToRESTConfig(cluster *apimgmtv3.Cluster, context *config.ScaledContext, secretLister v1.SecretLister) (*rest.Config, error) {
+// ToRESTConfig generates a rest.Config for a given cluster.
+// If tryReconnecting is true, the dialer used for connecting this rest.Config will block
+// and retry connecting to the cluster for ~30s if the connection is not available,
+// otherwise return immediately
+func ToRESTConfig(cluster *apimgmtv3.Cluster, context *config.ScaledContext, secretLister v1.SecretLister, tryReconnecting bool) (*rest.Config, error) {
```
[the remark this answers is not in the document]
Comment (moio) [reply pkg/clustermanager/manager.go]: I like that and will actually take it as a habit. Thanks!
Addressed as e2851cc31

File: pkg/clustermanager/manager.go [hunk 3]
```diff
@@ -467,12 +467,20 @@ func (m *Manager) APIExtClient(_ *types.APIContext, _ types.StorageContext) (cli
 
 // UserContextNoControllers accepts a cluster name and returns a client for that cluster,
 // no controllers are started for that cluster in the process.
+// Note it will block retrying to connect to the cluster for ~30 seconds before returning
+// in case the cluster connection fails.
 func (m *Manager) UserContextNoControllers(clusterName string) (*config.UserContext, error) {
```
[the remark this answers is not in the document]
Comment (moio) [reply pkg/clustermanager/manager.go:472]: Totally agreed both on the idea of a fluent API and the consideration this is best left for a follow-up refactoring PR once work on this is finalized.

I am resolving this conversation for now.

File: pkg/clustermanager/manager.go [hunk 4]
```diff
@@ -467,12 +467,20 @@ func (m *Manager) APIExtClient(_ *types.APIContext, _ types.StorageContext) (cli
...
+	return m.UserContextNoControllersReconnecting(clusterName, true)
+}
+
+// UserContextNoControllersReconnecting works like UserContextNoControllers if tryReconnecting is true.
+// Otherwise, it will return an error immediately if the cluster connection fails.
+func (m *Manager) UserContextNoControllersReconnecting(clusterName string, tryReconnecting bool) (*config.UserContext, error) {
```
[the remark this answers is not in the document]
Comment (moio) [reply pkg/clustermanager/manager.go]: I count 20 call sites that need to retain the current behavior, and 2 call sites where we want change it (perhaps they will grow a bit in the future).

The trade-off to strike is: verbosity here (in the form of the alias func) vs clarity in most of the calling locations (what does this extra `true` do?)

With this patch intent was to highlight the places where we are changing things, and leave alone places where we are not really changing anything. As you mentioned to lower review noise, also potentially make cherry-picking easier.

But the final word is really on maintnainers I believe, and I am totally fine reversing the decision. What would be your preference?

File: pkg/types/config/dialer/dialer.go [hunk 5]
```diff
@@ -8,7 +8,11 @@ import (
 type Dialer func(ctx context.Context, network, address string) (net.Conn, error)
 
 type Factory interface {
-	ClusterDialer(clusterName string) (Dialer, error)
+	// ClusterDialer returns a dialer that can be used to connect to a cluster's Kubernetes API
+	// Note that the dialer may or may not use a remotedialer tunnel
+	// If retryOnError is true, the dialer will retry for ~30s in case it cannot connect,
+	// otherwise return immediately
+	ClusterDialer(clusterName string, retryOnError bool) (Dialer, error)
```
[the remark this answers is not in the document]
Comment (moio) [reply pkg/types/config/dialer/dialer.go:15]: Agreed, this should go to a refactoring PR. Resolving this conversation for now

File: pkg/controllers/managementuser/healthsyncer/healthsyncer.go [hunk 6]
```diff
@@ -130,6 +127,19 @@ func (h *HealthSyncer) getComponentStatus(cluster *v3.Cluster) error {
...
 }
 
+// IsAPIUp checks if the Kubernetes API server is up and etcd is available.
+// It gets a namespace from the API, even if not found, it means the API is up.
+// It returns nil if the API is up, otherwise it returns an error.
+func IsAPIUp(ctx context.Context, k8s kubernetes.Interface) error {
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	if _, err := k8s.CoreV1().Namespaces().Get(ctx, "kube-system", metav1.GetOptions{}); err != nil && !apierrors.IsNotFound(err) {
```
[the remark this answers is not in the document]
Comment (moio) [reply pkg/controllers/managementuser/healthsyncer/healthsyncer.go]: Easier is in the eye of the beholder I guess!

I tentatively pushed the approach you suggested (or at least what I hope I got correctly) as 666c2ff91.

I do not have any personal preference - on one hand using simple interfaces more and complicated mock mechanisms less is better, and I agree test code is now easier to follow. On the other hand, it does push a bit of extra complexity to `IsAPIUp` call sites, and those could grow a bit in count in the future, and test code is more verbose.

As you prefer. Please let me know if you actually like it better or I can also just remove it.

**Comment by moio** [comment]
Thanks for the reviews! Minor details still up to discussion.

I will merge as soon as either @erkki or some other user can confirm this helps out there in the field.

**Comment by erkki** [comment]
> Thanks for the reviews! Minor details still up to discussion.
> 
> I will merge as soon as either @erkki or some other user can confirm this helps out there in the field.

We in fact just deployed the pre-updated version in the field yesterday, will rebase on top of the latest changes and give it a few days of running in the wild.

**Review by Oats87 — APPROVED** [review]

**Review by jakefhyde — APPROVED** [review]

**Comment by erkki** [comment]
> > Thanks for the reviews! Minor details still up to discussion.
> > I will merge as soon as either @erkki or some other user can confirm this helps out there in the field.
> 
> We in fact just deployed the pre-updated version in the field yesterday, will rebase on top of the latest changes and give it a few days of running in the wild.

We have now been running the patch-set with all of our existing patches reverted (including bumping of controller numbers) and things are looking sane and stable. It took us longer to isolate as we were running into some unrelated issues (for which we will send a new report) but this patch looks stable -- job times are sane and seemingly no queue blockages.

**Pull request merged**